operator/cnp_derivative.go

```diff
@@ -0,0 +1,240 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cilium/cilium/pkg/controller"
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	clientset "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy/api"
+	_ "github.com/cilium/cilium/pkg/policy/groups"
+
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	cnpKindName = "derivative"
+	parentCNP   = "io.cilium.network.policy.parent.uuid"
+	cnpKindKey  = "io.cilium.network.policy.kind"
+
+	// maxNumberOfAttempts Number of times that try to retrieve a information from a cloud provider.
+	maxNumberOfAttempts = 5
+	// SleepDuration time that sleep in case that can't retrieve information from a cloud provider.
+	sleepDuration = 5 * time.Second
+)
+
+var (
+	controllerManager     = controller.NewManager()
+	globalK8sClient       *clientset.Clientset
+	k8sMutex              = lock.Mutex{}
+	blockOwnerDeletionPtr = true
+)
+
+func getDerivativeName(cnp *cilium_v2.CiliumNetworkPolicy) string {
+	return fmt.Sprintf(
+		"%s-togroups-%s",
+		cnp.GetObjectMeta().GetName(),
+		cnp.GetObjectMeta().GetUID())
+}
+
+// createDerivativeCNP will return a new CNP based on the given rule.
+func createDerivativeCNP(cnp *cilium_v2.CiliumNetworkPolicy) (*cilium_v2.CiliumNetworkPolicy, error) {
+	// CNP informer may provide a CNP object without APIVersion or Kind.
+	// Setting manually to make sure that the derivative policy works ok.
+	derivativeCNP := &cilium_v2.CiliumNetworkPolicy{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      getDerivativeName(cnp),
+			Namespace: cnp.ObjectMeta.Namespace,
+			OwnerReferences: []v1.OwnerReference{{
+				APIVersion:         cilium_v2.SchemeGroupVersion.String(),
+				Kind:               cilium_v2.CNPKindDefinition,
+				Name:               cnp.ObjectMeta.Name,
+				UID:                cnp.ObjectMeta.UID,
+				BlockOwnerDeletion: &blockOwnerDeletionPtr,
+			}},
+			Labels: map[string]string{
+				parentCNP:  string(cnp.ObjectMeta.UID),
+				cnpKindKey: cnpKindName,
+			},
+		},
+	}
+
+	rules, err := cnp.Parse()
+	if err != nil {
+		return nil, fmt.Errorf("Cannot parse policies: %s", err)
+	}
+
+	derivativeCNP.Specs = make(api.Rules, len(rules))
+	for i, rule := range rules {
+		if rule.RequiresDerivative() {
+			derivativeCNP.Specs[i] = denyEgressRule()
+		}
+	}
+
+	for i, rule := range rules {
+		if !rule.RequiresDerivative() {
+			derivativeCNP.Specs[i] = rule
+			continue
+		}
+		newRule, err := rule.CreateDerivative()
```

Comment:
One of the commits is titled "Move CNP specific handling into operator". I wonder if we should also move `CreateDerivative` to this file. It seems to only call `CreateDerivative` on EgressRule and then just iterates through modifying public fields in `api.Rule` and `api.EgressRule` for both ToGroups and ToServices. This way all derivative code is in the operator (I don't think cilium ever checks whether something is a derivative).

Reply:
I think it's better to leave derivative creation in the pkg dir, this way it can be reused elsewhere if needed.